crates/wasmtime/src/store.rs

```diff
@@ -1945,29 +1943,28 @@ unsafe impl<T> wasmtime_runtime::Store for StoreInner<T> {
     fn new_epoch(&mut self) -> Result<u64, anyhow::Error> {
         // Temporarily take the configured behavior to avoid mutably borrowing
         // multiple times.
-        let mut behavior = std::mem::take(&mut self.epoch_deadline_behavior);
+        let mut behavior = self.epoch_deadline_behavior.take();
         let delta_result = match &mut behavior {
-            EpochDeadline::Trap => Err(Trap::Interrupt.into()),
-            EpochDeadline::Callback(callback) => {
-                let delta = callback((&mut *self).as_context_mut())?;
+            None => Err(Trap::Interrupt.into()),
+            Some(callback) => {
+                let delta = match callback((&mut *self).as_context_mut()) {
+                    // Note: If the callback returns an error, we don't call it again.
+                    DeadlineResult::Trap(err) => return Err(err),
+                    DeadlineResult::Continue(delta) => delta,
+                    #[cfg(feature = "async")]
+                    DeadlineResult::Yield(delta) => {
+                        // Do the async yield. May return a trap if future was
+                        // canceled while we're yielded.
+                        self.async_yield_impl()?;
+                        delta
+                    }
```

Comment:
In this case I believe that `async_yield_impl` will panic internally if it's not configured right but technically that method's panic should be allowable to be a `debug_assert!` or an `unwrap_unchecked` or something like that. I think it'd be good to have a dedicated assert here for the use case that this covers where `Yield` is returned but async isn't enabled or configured.